source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,242 @@
+                                                                                        const std::string& service,
+                                                                                        const bool local) {
+  return getConnectionPeer(&connection, service, local);
+}
+
+
+const std::string ExtAuthzCheckRequestGenerator::getProtocolStr(const Envoy::Http::Protocol& p) {
+
+  switch(p) {
+    case Envoy::Http::Protocol::Http10:
+      return std::string("Http1.0");
+    case Envoy::Http::Protocol::Http11:
+      return std::string("Http1.1");
+    case Envoy::Http::Protocol::Http2:
+      return std::string("Http2");
+    default:
+      break;
+  }
+  return std::string("unknown");
+}
+
+Envoy::Http::HeaderMap::Iterate ExtAuthzCheckRequestGenerator::fillHttpHeaders(const Envoy::Http::HeaderEntry &e,
+                                                                               void *ctx) {
+  ::google::protobuf::Map< ::std::string, ::std::string >* mhdrs = static_cast<::google::protobuf::Map< ::std::string, ::std::string >*>(ctx);
+  (*mhdrs)[std::string(e.key().c_str(), e.key().size())] = std::string(e.value().c_str(), e.value().size());
+  return Envoy::Http::HeaderMap::Iterate::Continue;
+}
+
+std::string ExtAuthzCheckRequestGenerator::getHeaderStr(const Envoy::Http::HeaderEntry *entry) {
+  if (entry) {
+    return std::string(entry->value().c_str(), entry->value().size());
+  }
+  return "";
+}
+
+std::unique_ptr<AttributeContext_Request> ExtAuthzCheckRequestGenerator::getHttpRequest(const Envoy::Http::StreamDecoderFilterCallbacks* callbacks,
+                                                                                        const Envoy::Http::HeaderMap &headers) {
+
+
+  AttributeContext_HttpRequest *httpreq = new AttributeContext_HttpRequest();
+  ASSERT(httpreq);
+
+  // Set id
+  // The streamId is not qualified as a const. Although it is as it does not modify the object.
+  Envoy::Http::StreamDecoderFilterCallbacks *sdfc = const_cast<Envoy::Http::StreamDecoderFilterCallbacks *>(callbacks);
+  httpreq->set_id(std::to_string(sdfc->streamId()));
+
+  // Set method
+  httpreq->set_method(std::move(getHeaderStr(headers.Method())));
+  // Set path
+  httpreq->set_path(std::move(getHeaderStr(headers.Path())));
+  // Set host
+  httpreq->set_host(std::move(getHeaderStr(headers.Host())));
+  // Set scheme
+  httpreq->set_scheme(std::move(getHeaderStr(headers.Scheme())));
+
+  // Set size
+  // need to convert to google buffer 64t;
+  httpreq->set_size(sdfc->requestInfo().bytesReceived());
+
+  // Set protocol
+  httpreq->set_protocol(getProtocolStr(sdfc->requestInfo().protocol().value()));
+
+  // Fill in the headers
+  auto* mhdrs = httpreq->mutable_headers();
+  headers.iterate(fillHttpHeaders, mhdrs);
+
+  std::unique_ptr<AttributeContext_Request> req = std::unique_ptr<AttributeContext_Request>{new AttributeContext_Request()};
+  ASSERT(req);
```

Comment:
Very brief drive-by nits here:
1. Don't need to `ASSERT` on a basic memory allocation, we assume memory allocations always succeed or the world ends in Envoy.
2. Should write this as `auto req = std::make_unique<Attribute_Request>();` for C++14.